### ISSUE TITLE
Provide a library for simple valued named constants

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5383]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#5383|
|Type|enhancement|
|Created|2011-11-13 19:33:05Z|
|Branch|https://github.com/twisted/twisted/tree/valued-constants-5383|

[#4671](https://github.com/twisted/twisted/issues/4671) proposed adding a lot of new symbolic constant features to a helper somewhere in Twisted. This is for the kind of symbolic constant that has an associated value (perhaps unique?  perhaps not?  perhaps those are both valueable use cases), assigned either manually or automatically from a monotonically increasing sequences of integers.

This would be useful in the HTTP implementation, in the IRC implementation, in the telnet implementation, and probably in lots more places.

<details><summary>Searchable metadata</summary>

```
trac-id__5383 5383
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_valued_constants_5383 branches/valued-constants-5383
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1321212785000000 1321212785000000
changetime__1323553163000000 1323553163000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [33282]) Branching to 'valued-constants-5383'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun removed owner|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/valued-constants-5383)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst set owner to @itamarst<br>@itamarst set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst set owner to @exarkun<br>@itamarst set status to `new`|
|-|-|

  1. `_ConstantsContainer` still has references to `NamedConstants` and `Names`. Likewise `_EnumerantsInitializer`.
  1. I'd move `NamedConstant` to be next to `Names`, rather than off elsewhere in the code. 
  1. `lookupByValue` should document, and test, that if there are multiple constants with same value, the first one will be returned.

Once you fix those, feel free to merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [33329]) document and test value constants with duplicate values

refs [#5383](https://github.com/twisted/twisted/issues/5383)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [33330]) Merge valued-constants-5383 

Author: exarkun
Reviewer: itamar
Fixes: [#5383](https://github.com/twisted/twisted/issues/5383)

Add `twisted.python.constants.Values`, a utility for defining named constants with
associated values.
